src/parser/input_parser.cpp

```diff
@@ -0,0 +1,91 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Andres Noetzli
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * The interface for parsing an input with a parser.
+ */
+
+#include "parser/input_parser.h"
+
+#include "base/output.h"
+#include "parser/input.h"
+#include "parser/parser.h"
+#include "smt/command.h"
+
+namespace cvc5 {
+namespace parser {
+
+Command* InputParser::nextCommand()
+{
+  Debug("parser") << "nextCommand()" << std::endl;
+  Command* cmd = nullptr;
+  if (d_state->hasCommand())
+  {
+    cmd = d_state->getNextCommand();
+    d_state->setDone(cmd == nullptr);
```

Comment:
This feels like `d_state` should implement this logic itself. In particular, as the logic is repeated below.